### PR TITLE
scripts/brp-strip: Rectify parallel stripping logic

### PR DESCRIPTION
**Pull request by sshedi**
brp-strip script runs strip command on deliverables paralley and if
deliverables are hard linked inside buildroot, it will create
contention.

One good example for such package is git.
https://github.com/vmware/photon/blob/master/SPECS/git/git.spec

```
Sample output:
$ rpm -ql git | grep libexec | xargs ls -li
668153 -rwxr-xr-x  137 root root 3401056 Aug  2 08:30 /usr/libexec/git-core/git
668153 -rwxr-xr-x  137 root root 3401056 Aug  2 08:30 /usr/libexec/git-core/git-add
787238 -rwxr-xr-x    1 root root   47770 Aug  2 08:30 /usr/libexec/git-core/git-add--interactive
668153 -rwxr-xr-x  137 root root 3401056 Aug  2 08:30 /usr/libexec/git-core/git-am
```

Hence we need to identify and discard hardlinks and use a single
instance of the elf file to run strip command.

This is done by running a crafted command to get inode of files and
removing duplicate inodes from the result.

RH bug link:
https://bugzilla.redhat.com/show_bug.cgi?id=1959049

Signed-off-by: Shreenidhi Shedi <sshedi@vmware.com>
Co-authored-by: Dweep Advani <dadvani@vmware.com>

### CODE REVIEW

**Comment by pmatilai** [comment]
This causes indigestion on `file`, the following is present in the test-suite log multiple times:
```
+Usage: file [-bcCdEhikLlNnprsSvzZ0] [--apple] [--extension] [--mime-encoding]
+            [--mime-type] [-e <testname>] [-F <separator>]  [-f <namefile>]
+            [-m <magicfiles>] [-P <parameter=value>] [--exclude-quiet]
+            <file> ...
+       file -C [-m <magicfiles>]
+       file [--help]
```

So something's not right there, and I think we'll need to find some nicer solution to the problem, as all our parallel brp-scripts are similarly affected, even if nobody has encountered the issue elsewhere yet. Thanks for the identifying the issue!

**Comment by sshedi** [comment]
@pmatilai Thanks for the response, I will check the failure logs further. Glad that you liked the idea.

**Comment by ffesti** [comment]
This patch has several obvious issues:
- Still runs everything in parallel when the `sort -u` obviously needs to be run as single process
- ARG0 disappeared for no reason silently dropping one file from processing
- number of arguments changed for no reason and without mentioning it in the commit message 
  - this basically disables the parallel processing and is the reason this change "fixes" the issue
- the backslash escape for the `!` of the `find` command disappeared for no reason which is probably causing the failure

and that's without me even trying to run it...

**Comment by ffesti** [comment]
> Thanks @ffesti for the inputs. Find my responses inline.
> 
> > This patch has several obvious issues:
> > [sshedi]: Yes indeed. I'm open for inputs
> 
> ```
> > Still runs everything in parallel when the sort -u obviously needs to be run as single process
> ```
> 
> [sshedi]: Yes, agreed. But since we are processing in batch of files and if elf is stripped, sed command will eliminate it. So sort command won't get files which are already stripped and if a file is stripped once, it won't reach till sort command in pipeline.

This is still prone to race conditions. Especially as the front part of the pipe line will be able to process much more files more quickly than the end.

> > number of arguments changed for no reason and without mentioning it in the commit message
> ```
> 
> [sshedi]: I believe you are talking about `-n256` option here. We can pass up to 2MB size in argument, so we can use 256 here. Considering PATH_MAX is 4096, 4096 * 256 = 1MB so there is still 1MB left. I will mention this in commit message.
> 
> ```
> ➜ $ getconf ARG_MAX
> 2097152 (2MB)
> ```
> 
> ```
>     > this basically disables the parallel processing and is the reason this change "fixes" the issue
> ```
> 
> [sshedi]: Can you please elaborate on this? How does this disable parallel processing?

Well, you have at most `number of files` / `maxargs` processes. If `maxargs` is bigger then the number of files there is only one process. `-n32` was chosen as a compromise between reducing the overhead of starting new processes and distributing the work load evenly over as much processors as possible. If you are testing with less than 256 files you are not testing parallel processing at all. The maximum allowable size of the arguments is not te limiting factor here.

> ```
> > the backslash escape for the ! of the find command disappeared for no reason which is probably causing the failure
> ```
> 
> [sshedi]: the ! is redundant in this case and it works perfectly fine without it, no need to escape it because we running in batch mode and not interactive mode. Correct me if I'm wrong here.

turns out the escaping is indeed not needed. This still makes me a bit uneasy.
 
> > and that's without me even trying to run it...
> > [sshedi]: Please give it a run :)
> 
> Thanks for reviewing.

**Comment by sshedi** [comment]
> This is still prone to race conditions. Especially as the front part of the pipe line will be able to process much more files more quickly than the end.
So if I have understood it right, using `-n32` should solve both the issues you pointed out.

**Comment by sshedi** [comment]
@ffesti how does this version look to you? Yes you were right about sort running as a single independent process.

**Comment by sshedi** [comment]
@pmatilai & @ffesti can you please take a look at the updated changes?

File: scripts/brp-strip [hunk 1]
```diff
@@ -7,11 +7,43 @@ fi
...
+# Only operate on non-stripped binaries
+# Get inode & file names to help eliminate hard links (%i %n)
+tmp_fn="$(mktemp)"
+find "$RPM_BUILD_ROOT" -type f \
+  ! -regex "${RPM_BUILD_ROOT}/?usr/lib/debug/.*" \
+  ! -name "*.go" -print0 | \
+  xargs -0 -r -P${NCPUS} -n${MAX_ARGS} sh -c "file \"\$@\" | \
+  sed -n -e 's/^\(.*\):[[:space:]]*ELF.*, not stripped.*/\1/p' | \
+  grep -v 'no machine' | \
+  xargs -r stat -c '%i %n'" ARG0 > "${tmp_fn}"
```
Comment (ffesti) [inline scripts/brp-strip:36]: Isn't this overwriting `${tmp_fn}` from multiple processes? Using `>>` for the redirection should solve this. Still having this tested with a significant number of binaries would make be sleep better.

**Comment by ffesti** [comment]
Having three distinct steps reduces the parallelism a bit. May be this can be avoided using a named pipe instead of a temp file. But I am not sure that's worth the added complexity. So this is more for future reference...

File: scripts/brp-strip (hunk 1 above)
Comment (sshedi) [reply scripts/brp-strip:36]: Yeah, I got this thought. Surprisingly it doesn't and I tested this with a random find command with xargs like the way it's used here. Anyway, I will do this change.

**Review by sshedi — COMMENTED** [review]
Having 3 distinct steps seems a bit costly but it's needed.

Assume we first 32 lines in find command results are individual binaries but 33rd entry is a hard linked file. And sort command will act on batch of 32 lines, so there is a chance that in the sort will give same binary to strip command at same time. So, sort has to work on the whole list of files at once.

And using a named pipe is an excellent idea and huge performance gain compared to using a temp file. Thanks for the input 👍🏻

**Comment by sshedi** [comment]
Sorry, I did something wrong with my git branch and I have created a new PR
https://github.com/rpm-software-management/rpm/pull/2161

**Comment by DemiMarie** [comment]
Can you make sure this works with spaces, newlines, and other weird characters in file names?

**Comment by sshedi** [comment]
Hi @DemiMarie I think there is a test that's testing what your asking, I have seen it failing in one my change revision.

So I think we are good.

**Comment by sshedi** [comment]
Something like this: https://github.com/rpm-software-management/rpm/blob/21c34bdd61c14f8044fcc5c0be90b1ec55eef2df/tests/data/SPECS/prefixtest.spec#L35

**Pull request closed without merging**